### PR TITLE
[REVIEW] rolling support for durations types

### PR DESCRIPTION
**Pull request by karthikeyann**
addresses part of https://github.com/rapidsai/cudf/issues/5272

~- [x] typed tests to include the new types~
- [x] add rolling, grouped_rolling support for duration types
~- add duration types to fixed_width types and chrono types~
- add duration_scalar fixes to multiple test files (fix for seg fault)
~- add explicit typecasting from `int` to `TypeParam` for fixed_width types tests.~
~- add specialization of explict typecast from `initializer_list<int>` to duration types in `fixed_width_column_wrapper` to ease the unit tests.~

### CODE REVIEW

File: cpp/include/cudf/detail/utilities/transform_unary_functions.cuh [hunk 1]
```diff
@@ -30,6 +30,21 @@
...
 
 namespace cudf {
+
+/**
+ * @brief Functor to static_cast convertible types.
+ *
+ * @tparam result_type Target type to be typecasted to.
+ */
+template <typename ResultType>
+struct typecaster {
```
Comment (sriramch) [inline cpp/include/cudf/detail/utilities/transform_unary_functions.cuh:40]: this assumes that `ElementType` is convertible to `ResultType` which may not be case always.

File: cpp/tests/utilities/column_wrapper.hpp [hunk 2]
```diff
@@ -112,7 +113,10 @@ template <typename Element, typename InputIterator>
 rmm::device_buffer make_elements(InputIterator begin, InputIterator end)
 {
   static_assert(cudf::is_fixed_width<Element>(), "Unexpected non-fixed width type.");
-  thrust::host_vector<Element> elements(begin, end);
+  auto begin_t =
+    thrust::make_transform_iterator(begin, cudf::typecaster<Element>{});  // explicit typecast
```
Comment (sriramch) [inline cpp/tests/utilities/column_wrapper.hpp:117]: the iterator can produce types that may not convertible to the target type `Element`. hence, you may have to distinguish between cases where the types are convertible and where they do not. for ones that can't be converted, you may have to check on a case by case basis if the iterators value type can be converted to the target type.

please see [here](https://github.com/rapidsai/cudf/blob/8fca5d5073fce281b3cf979fac21694b0747e83b/cpp/tests/utilities/column_wrapper.hpp#L113) and [here](https://github.com/rapidsai/cudf/blob/8fca5d5073fce281b3cf979fac21694b0747e83b/cpp/tests/utilities/column_wrapper.hpp#L157) for one possible approach.

File: cpp/include/cudf/detail/utilities/transform_unary_functions.cuh (hunk 1 above)
Comment (karthikeyann) [reply cpp/include/cudf/detail/utilities/transform_unary_functions.cuh:40]: That will create a compilation error. not a runtime error.

File: cpp/tests/utilities/column_wrapper.hpp (hunk 2 above)
Comment (karthikeyann) [reply cpp/tests/utilities/column_wrapper.hpp:117]: Here, explicit non-convertible types won't compile anyway.
explicit and implicit type conversion should have the same generated assembly finally. no need for specialization.

Comment (sriramch) [reply cpp/tests/utilities/column_wrapper.hpp:117]: there are tests that may create fixed width types with numeric types, which may fail if they aren't convertible. it is being tested today and we want to support that for the new type as well.

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: cpp/tests/utilities/column_wrapper.hpp (hunk 2 above)
Comment (karthikeyann) [reply cpp/tests/utilities/column_wrapper.hpp:123]: removed the general iterator explicit typecast. 
Added specialization for initializer_list<int> to chrono types.

File: cpp/tests/utilities/column_wrapper.hpp [hunk 3]
```diff
@@ -206,7 +206,7 @@ auto make_chars_and_offsets(StringsIterator begin, StringsIterator end, Validity
  *
  * @tparam Element The fixed-width element type
  **/
-template <typename ElementTo>
+template <typename ElementTo, typename ConstructFrom = ElementTo>
```
Comment (karthikeyann) [inline cpp/tests/utilities/column_wrapper.hpp:209]: @jrhemstad  This implement the review comment https://github.com/rapidsai/cudf/pull/5394#discussion_r443905361 
please review.

**Comment by karthikeyann** [comment]
> This all looks good, @karthikeyann . My main observation is that nearly all of the _many_ changed lines are simply adding the same second template parameter to instantiations of `fixed_width_column_wrapper`. Careful use of a `using` alias in each of the tests could guard against needing to change so many lines in the future. Or possibly an appropriate default type for the parameter.

I thought about use of `using` alias, but skipped it. Recalling what @jrhemstad said here https://github.com/rapidsai/cudf/pull/5394#discussion_r438512267 , an _explicit_ parameter at the initializer list would be good idea for the caller to know that explicit type conversion is taking place. so went ahead with many line changes.

if `using` alias in each translation unit is still considered explicit, I would code it that way.

File: cpp/include/cudf/utilities/traits.hpp [hunk 4]
```diff
@@ -63,6 +63,18 @@ template <typename L, typename R>
 struct is_equality_comparable_impl<L, R, void_t<equality_comparable<L, R>>> : std::true_type {
 };
 
+template <class From, class To>
+struct is_explicitly_convertible {
```
Comment (jrhemstad) [inline cpp/include/cudf/utilities/traits.hpp:67]: Since these aren't standard traits, you can just make them `constexpr` functions.

**Comment by jrhemstad** [comment]
> This all looks good, @karthikeyann . My main observation is that nearly all of the _many_ changed lines are simply adding the same second template parameter to instantiations of `fixed_width_column_wrapper`. Careful use of a `using` alias in each of the tests could guard against needing to change so many lines in the future. Or possibly an appropriate default type for the parameter.

As @karthikeyann said, the new changes to `fixed_width_column_wrapper` are performing a lot of conversions that would otherwise need to be explicit. I wasn't comfortable with making it totally implicit and hidden from the user that these conversions are taking place. Implicit conversions are sources of a lot of confusing and difficult to triage behaviors. 

The second explicit template argument forces awareness of a conversion taking place from the type of the init list to whatever the `fixed_width_column`s type is.

File: cpp/tests/utilities/column_wrapper.hpp [hunk 5]
```diff
@@ -291,10 +296,63 @@ class fixed_width_column_wrapper : public detail::column_wrapper {
...
+   *
+   * @tparam ValidityIterator Dereferencing a ValidityIterator must be
+   * convertible to `bool`
+   * @param elements The list of elements
+   * @param v The beginning of the sequence of validity indicators
+   **/
+  template <typename ValidityIterator,
+            typename ElementFrom,
+            typename std::enable_if_t<!is_explicitly_convertible<ElementFrom>()>* = nullptr>
+  fixed_width_column_wrapper(std::initializer_list<ElementFrom> elements, ValidityIterator v)
```
[the remark this answers is not in the document]
Comment (karthikeyann) [reply cpp/tests/utilities/column_wrapper.hpp:341]: Right. I will remove these once PR #5394 is merged.

[the remark this answers is not in the document]
Comment (karthikeyann) [reply cpp/tests/utilities/column_wrapper.hpp]: updated.

File: cpp/tests/rolling/rolling_test.cpp [hunk 6]
```diff
@@ -581,7 +581,7 @@ TYPED_TEST(RollingTest, ZeroWindow)
 {
   size_type num_rows = 1000;
 
-  std::vector<TypeParam> col_data(num_rows, 1);
+  std::vector<TypeParam> col_data(num_rows, TypeParam(1));
```
Comment (sriramch) [inline cpp/tests/rolling/rolling_test.cpp:584]: wouldn't you have to include the duration types for this typed test for these changes to be necessary?

File: cpp/tests/grouped_rolling/grouped_rolling_test.cpp [hunk 7]
```diff
@@ -1214,7 +1214,7 @@ TYPED_TEST(GroupedTimeRangeRollingTest,
 TYPED_TEST(GroupedTimeRangeRollingTest, SimplePartitionedStaticWindowsWithNoGroupingKeys)
 {
   const size_type DATA_SIZE{static_cast<size_type>(6)};
-  const std::vector<TypeParam> col_data(DATA_SIZE, 1);
+  const std::vector<TypeParam> col_data(DATA_SIZE, TypeParam(1));
```
Comment (sriramch) [inline cpp/tests/grouped_rolling/grouped_rolling_test.cpp:1217]: wouldn't you have to include the duration types for this typed test (`GroupedTimeRangeRollingTest`) for these changes to be necessary?

File: cpp/tests/grouped_rolling/grouped_rolling_test.cpp [hunk 8]
```diff
@@ -552,11 +552,11 @@ TYPED_TEST(GroupedRollingTest, SimplePartitionedStaticWindowsWithGroupKeys)
   // 2 grouping keys, with effectively 3 groups of at most 4 rows each:
   //   1. key_0 {0, 0, 0, ...0}
   //   2. key_1 {0, 0, 0, 0, 1, 1, 1, 1, 2, 2}
-  std::vector<TypeParam> key_0_vec(DATA_SIZE, 0);
+  std::vector<TypeParam> key_0_vec(DATA_SIZE, TypeParam(0));
```
Comment (sriramch) [inline cpp/tests/grouped_rolling/grouped_rolling_test.cpp:555]: wouldn't you have to include the duration types for this typed test (`GroupedRollingTest`) for these changes to be necessary?

File: cpp/tests/rolling/rolling_test.cpp (hunk 6 above)
Comment (karthikeyann) [reply cpp/tests/rolling/rolling_test.cpp:584]: Merging removed it. will add it again.

File: cpp/tests/grouped_rolling/grouped_rolling_test.cpp (hunk 8 above)
Comment (karthikeyann) [reply cpp/tests/grouped_rolling/grouped_rolling_test.cpp:552]: added.

File: cpp/tests/rolling/rolling_test.cpp (hunk 6 above)
Comment (karthikeyann) [reply cpp/tests/rolling/rolling_test.cpp:581]: Added.

**Pull request merged**